Sources/Support/POSIX/ELFSupport.cpp

```diff
@@ -69,5 +71,93 @@ bool ELFSupport::MachineTypeToCPUType(uint32_t machineType, bool is64Bit,
   }
   return true;
 }
+
+bool ELFSupport::GetELFFileBuildID(std::string const &path, ByteVector &buildId) {
+  int fd = ::open(path.c_str(), O_RDONLY);
+  if (fd < 0)
+    return false;
+
+  Elf32_Ehdr e32hdr;
+  if (::pread(fd, &e32hdr, sizeof(e32hdr), 0) != sizeof(e32hdr) ||
+      e32hdr.e_ident[EI_MAG0] != ELFMAG0 || e32hdr.e_ident[EI_MAG1] != ELFMAG1 ||
+      e32hdr.e_ident[EI_MAG2] != ELFMAG2 || e32hdr.e_ident[EI_MAG3] != ELFMAG3) {
+    DS2LOG(Warning, "File \"%s\" is not valid ELF format", path.c_str());
+    ::close(fd);
+    return false;
+  }
+
+  bool result = false;
+  switch (e32hdr.e_ident[EI_CLASS]) {
+    case ELFCLASS32:
+      Elf32_Shdr s32hdr;
+      Elf32_Nhdr n32hdr;
+      result = ReadBuildID(fd, e32hdr, s32hdr, n32hdr, buildId);
+      break;
+
+    case ELFCLASS64:
+      Elf64_Ehdr e64hdr;
+      if (::pread(fd, &e64hdr, sizeof(e64hdr), 0) != sizeof(e64hdr))
+        break;
+
+      Elf64_Shdr s64hdr;
+      Elf64_Nhdr n64hdr;
+      result = ReadBuildID(fd, e64hdr, s64hdr, n64hdr, buildId);
+      break;
+
+    default:
+      DS2LOG(Warning, "File \"%s\" contains unsupported ELF class identifier %i",
+             path.c_str(), e32hdr.e_ident[EI_CLASS]);
+      break;
+  }
+
+  ::close(fd);
+
+  if (!result)
+    DS2LOG(Warning, "Failed to query build ID for ELF file \"%s\"", path.c_str());
+
+  return result;
+}
+
+template <class T_EHDR, class T_SHDR, class T_NHDR>
```

Comment:
```suggestion
template <typename ELFHeader, typename SectionHeader, typename NotesHeader>
```

Reply:
Ah ok, `typename` is the right keyword here. I found `struct` didn't work, swapped in `class` and forgot to go back to find the right answer.

Reply:
Quite so, `class` was overloaded back in the day, but the proper keyword is `typename` as we are specifying a type (which is what the `T_` prefix is as well).